client/src/templates/Challenges/quiz/show.tsx

```diff
@@ -111,147 +165,148 @@ class ShowQuiz extends Component<ShowQuizProps, ShowQuizState> {
       helpCategory
     });
     challengeMounted(challengeMeta.id);
-    this.container.current?.focus();
-  }
+    container.current?.focus();
+    // This effect should be run once on mount
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
 
-  componentDidUpdate(prevProps: ShowQuizProps): void {
-    const {
-      data: {
-        challengeNode: {
-          challenge: { title: prevTitle }
-        }
-      }
-    } = prevProps;
-    const {
-      challengeMounted,
-      data: {
-        challengeNode: {
-          challenge: { title: currentTitle, challengeType, helpCategory }
+  useEffect(() => {
+    updateChallengeMeta({
+      ...challengeMeta,
+      title,
+      challengeType,
+      helpCategory
+    });
+    challengeMounted(challengeMeta.id);
+  }, [
+    title,
+    challengeMeta,
+    challengeType,
+    helpCategory,
+    challengeMounted,
+    updateChallengeMeta
+  ]);
+
+  const handleAnswersCheck = () => {
+    const unansweredQuestions = quizData.reduce(
+      (accumulator, currentQuestion, currentIndex) => {
+        if (currentQuestion.selectedAnswer == null) {
+          accumulator.push(currentIndex + 1);
         }
+        return accumulator;
       },
-      pageContext: { challengeMeta },
-      updateChallengeMeta
-    } = this.props;
-    if (prevTitle !== currentTitle) {
-      updateChallengeMeta({
-        ...challengeMeta,
-        title: currentTitle,
-        challengeType,
-        helpCategory
-      });
-      challengeMounted(challengeMeta.id);
+      [] as number[]
+    );
+
+    if (unansweredQuestions.length) {
+      setErrorMessage(
+        t('learn.quiz.unanswered-questions', {
+          unansweredQuestions: unansweredQuestions.join(', ')
+        })
+      );
+      return;
     }
-  }
 
-  handleSubmit() {
-    console.log('handleSubmit');
-  }
+    validateAnswers();
+    setHasSubmitted(true);
+  };
 
-  render() {
-    const {
-      data: {
-        challengeNode: {
-          challenge: {
-            title,
-            // challengeType,
-            description,
-            superBlock,
-            block,
-            translationPending,
-            quizzes
-          }
-        }
-      },
-      // openCompletionModal,
-      openHelpModal,
-      pageContext: {
-        challengeMeta: { nextChallengePath, prevChallengePath }
-      },
-      t,
-      isChallengeCompleted
-    } = this.props;
+  const handleSubmitAndGo = () => {
+    openCompletionModal();
+  };
 
-    const blockNameTitle = `${t(
-      `intro:${superBlock}.blocks.${block}.title`
-    )} - ${title}`;
+  // Handle error message display on submit.
+  // We keep the logic in a useEffect hook in order to subscribe to the `correctAnswerCount` change
+  // as `correctAnswerCount` is a React state and its value is updated asynchronously,
+  useEffect(() => {
+    if (!hasSubmitted) {
+      return;
+    }
```

Comment:
`useEffect` feels like the wrong tool for this. We're using it to handle a click event, but in a roundabout way.

As I understand what's happening this is the sequence

1. User fills in the quiz (maybe partially)
2. If partially complete, we ask them to finish
3. If complete we validate
4. If it fails we update the state returned by `useQuiz` and do some logic in the show component i.e. call a function (let's call it `onSuccess`)
5. If it succeeds we again update the `useQuiz` state and do some different logic (`onFailure`, say) in the show component

Everything is fine, except 4. and 5. Reason being, we shouldn't have to look at the state to know to call `onSuccess`  i.e. checking `hasSubmitted`'s value and comparing `correctAnswerCount` to `quiz.length`. `onSuccess` should be called immediately in response to the click, rather than indirectly when the state changes.

If that makes sense, I think we can implement it by passing `onSuccess` and  `onFailure` into `useQuiz`, so that it can call them when `validateAnswers` is called.

What do you reckon?